crates/ethrpc/src/buffered.rs

```diff
@@ -84,23 +91,83 @@ where
+                    itertools::multiunzip(
+                        batch
+                            .into_iter()
+                            .filter(|(_, _, _, sender)| !sender.is_canceled())
+                            .map(|(id, request, trace_id, sender)| {
+                                ((id, request), trace_id, sender)
+                            }),
+                    );
                 match requests.len() {
                     0 => (),
                     1 => {
-                        let ((id, request), sender) = (requests.remove(0), senders.remove(0));
-                        let result = inner.send(id, request).await;
+                        let ((id, request), trace_id, sender) =
+                            (requests.remove(0), trace_ids.remove(0), senders.remove(0));
+                        let result = match (&request, trace_id) {
+                            (Call::MethodCall(_), Some(trace_id)) => {
+                                observe::request_id::set_task_local_storage(
+                                    trace_id,
+                                    inner.send(id, request),
+                                )
+                                .await
+                            }
+                            _ => inner.send(id, request).await,
+                        };
                         let _ = sender.send(result);
                     }
                     n => {
-                        let results = inner
-                            .send_batch(requests)
-                            .await
-                            .unwrap_or_else(|err| vec![Err(err); n]);
+                        // Group requests by trace_id(sorted), then group values by method
+                        // name(unsorted) with call idx values(sorted).
+                        let mut result_map: BTreeMap<String, HashMap<String, BTreeSet<usize>>> =
+                            BTreeMap::new();
+                        for (idx, ((_, call), trace_id)) in
+                            requests.iter().zip(trace_ids).enumerate()
+                        {
+                            if let Call::MethodCall(call) = call {
+                                let trace_id = trace_id.unwrap_or("-".to_string());
+                                let method_name = call.method.clone();
+                                result_map
+                                    .entry(trace_id)
+                                    .or_default()
+                                    .entry(method_name)
+                                    .or_default()
+                                    .insert(idx);
+                            }
+                        }
+                        // Produces the following format:
+                        // `1001:eth_call(0,2),eth_sendTransaction(4)|1002:eth_call(1,3)`
+                        let request_metadata = result_map
```

Comment:
Just noticed that this will result in tons of small allocations.
Could we use a `Vec` as a buffer and the [write!](https://doc.rust-lang.org/std/macro.write.html) macro to avoid that?
If you follow Felix' request to move that logic into a separate function that seems reasonable to me.

Reply:
Used `String::push_str` which does the same under the hood.